File: android/src/main/java/com/supermap/rnsupermap/JSSTOMPManager.java
package com.supermap.rnsupermap;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.supermap.messagequeue.STOMPManager;
import com.supermap.messagequeue.STOMPReceiver;
import com.supermap.messagequeue.STOMPSender;

import java.util.Calendar;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by Myself on 2017/5/16.
 */

public class JSSTOMPManager extends ReactContextBaseJavaModule {
    public static final String REACT_CLASS = "JSSTOMPManager";
    public static Map<String,STOMPManager> mManagerList=new HashMap<String,STOMPManager>();
    STOMPManager mManager;

    public JSSTOMPManager(ReactApplicationContext context){super(context);}

    public static String registerId(STOMPManager manager){
        if(!mManagerList.isEmpty()) {
            for(Map.Entry entry:mManagerList.entrySet()){
                if(manager.equals(entry.getValue())){
                    return (String)entry.getKey();
                }
            }
        }

        Calendar calendar=Calendar.getInstance();
        String id=Long.toString(calendar.getTimeInMillis());
        mManagerList.put(id,manager);
        return id;
    }

    @Override
    public String getName(){
        return REACT_CLASS;
    }

    @ReactMethod
    public void createObj(Promise promise){
        try{
            STOMPManager manager = new STOMPManager();
            String _STOMPManagerId = registerId(manager);

            WritableMap map = Arguments.createMap();
            map.putString("_STOMPManagerId",_STOMPManagerId);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void initializeLibrary(Promise promise){
        try{
            STOMPManager.initializeLibrary();
            promise.resolve("init");
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void shutdownLibrary(Promise promise){
        try{
            STOMPManager.shutdownLibrary();
            promise.resolve("shutdown");
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void newSender(String managerId,boolean  useTopic,String name,Promise promise){
        try{
            STOMPManager manager = mManagerList.get(managerId);
            STOMPSender sender = manager.newSender(useTopic,name);
            String senderId = JSSTOMPSender.registerId(sender);

            WritableMap map = Arguments.createMap();
            map.putString("STOMPSenderId",senderId);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void newReceiver(String managerId,boolean  useTopic,String name,String clientID,Promise promise){
        try{
            STOMPManager manager = mManagerList.get(managerId);
            STOMPReceiver receiver = manager.newReceiver(useTopic,name,clientID);
            String receiverId = JSSTOMPReceiver.registerId(receiver);

            WritableMap map = Arguments.createMap();
            map.putString("STOMPReceiverId",receiverId);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void connection(String managerId,String  URI,String name,String password,Promise promise){
        try{
            STOMPManager manager = mManagerList.get(managerId);
            boolean isConnection = manager.connection(URI,name,password);

            WritableMap map = Arguments.createMap();
            map.putBoolean("isConnection",isConnection);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void disconnection(String managerId,Promise promise){
        try{
            STOMPManager manager = mManagerList.get(managerId);
            manager.disconnection();

            promise.resolve("disconnection");
        }catch (Exception e){
            promise.reject(e);
        }
    }
}

File: ios/objNative/JSSTOMPManager.m
//
//  JSSTOMPManager.m
//  Supermap
//
//  Created by 王子豪 on 2017/5/15.
//  Copyright © 2017年 Facebook. All rights reserved.
//

#import "JSSTOMPManager.h"

#import "SuperMap/STOMPManager.h"
#import "JSObjManager.h"
@implementation JSSTOMPManager
RCT_EXPORT_MODULE();
RCT_REMAP_METHOD(createObj,createObjWithresolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        STOMPManager* manager = [[STOMPManager alloc]init];
        NSInteger nsKey = (NSInteger)manager;
        [JSObjManager addObj:manager];
        resolve(@{@"_STOMPManagerId":@(nsKey).stringValue});
    } @catch (NSException *exception) {
        reject(@"STOMPManager",@"create Obj expection",nil);
    }
}

RCT_REMAP_METHOD(initializeLibrary,initializeLibraryWithresolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        [STOMPManager initializeLibrary];
        resolve(@"initializeLibrary");
    } @catch (NSException *exception) {
        reject(@"STOMPManager",@"initialize Library expection",nil);
    }
}

RCT_REMAP_METHOD(shutdownLibrary,shutdownLibraryWithresolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        [STOMPManager shutdownLibrary];
        resolve(@"shutdown Library");
    } @catch (NSException *exception) {
        reject(@"STOMPManager",@"shutdown Library expection",nil);
    }
}

RCT_REMAP_METHOD(newSender,newSenderWithId:(NSString*)managerId useTopic:(BOOL)UseTopic name:(NSString*)name resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        STOMPManager* manager = [JSObjManager getObjWithKey:managerId];
        STOMPSender* sender = [manager newSender:UseTopic name:name];
        NSInteger nsKey = (NSInteger)sender;
        [JSObjManager addObj:sender];
        resolve(@{@"STOMPSenderId":@(nsKey).stringValue});
    } @catch (NSException *exception) {
        reject(@"STOMPManager",@"shutdown Library expection",nil);
    }
}

RCT_REMAP_METHOD(newReceiver,newReceiverWithId:(NSString*)managerId useTopic:(BOOL)UseTopic name:(NSString*)name clientID:(NSString*)clientID resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        STOMPManager* manager = [JSObjManager getObjWithKey:managerId];
        STOMPReceiver* receiver = [manager newReceiver:UseTopic name:name clientid:clientID];
        NSInteger nsKey = (NSInteger)receiver;
        [JSObjManager addObj:receiver];
        resolve(@{@"STOMPReceiverId":@(nsKey).stringValue});
    } @catch (NSException *exception) {
        reject(@"STOMPManager",@"shutdown Library expection",nil);
    }
}

RCT_REMAP_METHOD(connection,connectionWithId:(NSString*)managerId URI:(NSString*)URI name:(NSString*)name passWord:(NSString*)passWord resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        STOMPManager* manager = [JSObjManager getObjWithKey:managerId];
        BOOL isConnection = [manager connection:URI username:name password:passWord];
        NSNumber* num = [NSNumber numberWithBool:isConnection];
        resolve(@{@"isConnection":num});
    } @catch (NSException *exception) {
        reject(@"STOMPManager",@"connection expection",nil);
    }
}

RCT_REMAP_METHOD(disconnection,disconnectionWithId:(NSString*)managerId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        STOMPManager* manager = [JSObjManager getObjWithKey:managerId];
        [manager disConnection];
        resolve(@"disConnection");
    } @catch (NSException *exception) {
        reject(@"STOMPManager",@"disConnection expection",nil);
    }
}
@end
